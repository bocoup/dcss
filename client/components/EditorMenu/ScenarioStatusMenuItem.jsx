import React from 'react';
import PropTypes from 'prop-types';
import changeCase from 'change-case';
import { Dropdown, Menu, Icon } from 'semantic-ui-react';

let cachedStatusOptions = null;

class ScenarioStatusMenuItem extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            statusOptions: null
        };

        this.fetchStatusOptions();
    }

    async fetchStatusOptions() {
        const statusOptions = await (cachedStatusOptions ||
            (await fetch('/api/status')).json());

        statusOptions.forEach(
            statusOption =>
                (statusOption.name = changeCase.titleCase(statusOption.name))
        );

        cachedStatusOptions = statusOptions; // eslint-disable-line require-atomic-updates

        this.setState({ statusOptions });
    }

    getStatusDisplay({ asIcon = false }) {
        const status = this.state.statusOptions.find(
            status => status.id === this.props.status
        );

        const color = {
            1: 'pink',
            2: 'green',
            3: 'orange'
        }[status.id];

        return asIcon ? (
            <Icon corner="top right" name="check" color={color} />
        ) : (
            `${status.name} (${status.description})`
        );
    }

    render() {
        if (!this.state.statusOptions) {
            return null;
        }
        return (
            <Menu.Menu key="menu-item-scenario-status">
                <Dropdown
                    item
                    text={
                        <React.Fragment>
                            <Icon.Group style={{ marginRight: '0.5rem' }}>
                                <Icon name="pencil alternate" />
                                {this.getStatusDisplay({
                                    asIcon: true
                                })}
                            </Icon.Group>
                            {this.getStatusDisplay({
                                asIcon: false
                            })}
                        </React.Fragment>
                    }
                >
                    <Dropdown.Menu>
                        {this.state.statusOptions.map(
                            ({ id, name, description }) => (
                                <Dropdown.Item
                                    name="save-status"
                                    id={id}
                                    key={id}
                                    onClick={this.props.onClick}
                                >
                                    {name} ({description})
                                </Dropdown.Item>
                            )
                        )}
                    </Dropdown.Menu>
                </Dropdown>
            </Menu.Menu>
        );
    }
}

ScenarioStatusMenuItem.propTypes = {
    // scenario: PropTypes.object,
    status: PropTypes.number,
    onClick: PropTypes.func.isRequired
};

export default ScenarioStatusMenuItem;
