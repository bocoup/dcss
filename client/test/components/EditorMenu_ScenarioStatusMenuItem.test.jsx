import React from 'react';
import assert from 'assert';
import {
  fetchImplementation,
  mounter,
  reduxer,
  snapshotter,
  state,
} from '../bootstrap';
import { unmountComponentAtNode } from 'react-dom';
import { mount, render, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ScenarioStatusMenuItem from '../../components/EditorMenu/ScenarioStatusMenuItem.jsx';

const original = JSON.parse(JSON.stringify(state));
let container = null;
let commonProps = null;
let commonState = null;

beforeAll(() => {
  (window || global).fetch = jest.fn();
});

afterAll(() => {
  fetch.mockRestore();
});

beforeEach(() => {
  container = document.createElement('div');
  container.setAttribute('id', 'root');
  document.body.appendChild(container);

  fetchImplementation(fetch);

  const statusOptions = [
    { id: 1, name: 'draft', description: 'Visible only to author' },
    { id: 2, name: 'public', description: 'Visible to everyone' },
    { id: 3, name: 'private', description: 'Visible only to logged in users' },
  ];
  fetchImplementation(fetch, 200, statusOptions);

  commonProps = {
    history: {
      push() {},
    },
  };

  commonState = JSON.parse(JSON.stringify(original));
});

afterEach(() => {
  fetch.mockReset();
  unmountComponentAtNode(container);
  container.remove();
  container = null;
  commonProps = null;
  commonState = null;
});

test('ScenarioStatusMenuItem', () => {
  expect(ScenarioStatusMenuItem).toBeDefined();
});

test('Snapshot 1 1', async (done) => {
  const Component = ScenarioStatusMenuItem;

  const props = {
    ...commonProps,
    onChange() {},
    status: 1,
  };

  const state = {
    ...commonState,
  };

  const reduxed = reduxer(Component, props, state);
  const wrapper = mounter(reduxed);
  expect(snapshotter(reduxed)).toMatchSnapshot();
  expect(snapshotter(wrapper)).toMatchSnapshot();

  const component = wrapper.findWhere((n) => {
    return n.type() === Component;
  });
  expect(snapshotter(component)).toMatchSnapshot();

  done();
});

test('Snapshot 2 1', async (done) => {
  const Component = ScenarioStatusMenuItem;

  const props = {
    ...commonProps,
    onChange() {},
    status: 2,
  };

  const state = {
    ...commonState,
  };

  const reduxed = reduxer(Component, props, state);
  const wrapper = mounter(reduxed);
  expect(snapshotter(reduxed)).toMatchSnapshot();
  expect(snapshotter(wrapper)).toMatchSnapshot();

  const component = wrapper.findWhere((n) => {
    return n.type() === Component;
  });
  expect(snapshotter(component)).toMatchSnapshot();

  done();
});

test('Snapshot 3 1', async (done) => {
  const Component = ScenarioStatusMenuItem;

  const props = {
    ...commonProps,
    onChange() {},
    status: 3,
  };

  const state = {
    ...commonState,
  };

  const reduxed = reduxer(Component, props, state);
  const wrapper = mounter(reduxed);
  expect(snapshotter(reduxed)).toMatchSnapshot();
  expect(snapshotter(wrapper)).toMatchSnapshot();

  const component = wrapper.findWhere((n) => {
    return n.type() === Component;
  });
  expect(snapshotter(component)).toMatchSnapshot();

  done();
});

/*{INJECTION}*/

