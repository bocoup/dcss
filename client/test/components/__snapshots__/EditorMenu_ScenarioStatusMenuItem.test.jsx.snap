// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render 1 1 1`] = `<DocumentFragment />`;

exports[`Render 1 1 2`] = `
<DocumentFragment>
  <div
    class="menu"
  >
    <div
      aria-expanded="true"
      class="ui active visible item dropdown"
      name="save-status"
      role="listbox"
      tabindex="0"
    >
      <i
        class="icons em__icon-group-margin"
      >
        <i
          aria-hidden="true"
          class="pencil alternate icon"
        />
        <i
          aria-hidden="true"
          class="pink check top right corner icon"
        />
      </i>
      Draft (Visible only to author)
      <i
        aria-hidden="true"
        class="dropdown icon"
      />
      <div
        class="visible menu transition"
      >
        <div
          aria-checked="true"
          aria-selected="true"
          class="active selected item"
          role="option"
          style="pointer-events: all;"
        >
          <span
            class="text"
          >
            Draft (Visible only to author)
          </span>
        </div>
        <div
          aria-checked="false"
          aria-selected="false"
          class="item"
          role="option"
          style="pointer-events: all;"
        >
          <span
            class="text"
          >
            Public (Visible to everyone)
          </span>
        </div>
        <div
          aria-checked="false"
          aria-selected="false"
          class="item"
          role="option"
          style="pointer-events: all;"
        >
          <span
            class="text"
          >
            Private (Visible only to logged in users)
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Render 1 1 3`] = `
<DocumentFragment>
  <div
    class="menu"
  >
    <div
      aria-expanded="false"
      class="ui item dropdown"
      name="save-status"
      role="listbox"
      tabindex="0"
    >
      <i
        class="icons em__icon-group-margin"
      >
        <i
          aria-hidden="true"
          class="pencil alternate icon"
        />
        <i
          aria-hidden="true"
          class="pink check top right corner icon"
        />
      </i>
      Draft (Visible only to author)
      <i
        aria-hidden="true"
        class="dropdown icon"
      />
      <div
        class="menu transition"
      >
        <div
          aria-checked="true"
          aria-selected="false"
          class="active item"
          role="option"
          style="pointer-events: all;"
        >
          <span
            class="text"
          >
            Draft (Visible only to author)
          </span>
        </div>
        <div
          aria-checked="false"
          aria-selected="false"
          class="item"
          role="option"
          style="pointer-events: all;"
        >
          <span
            class="text"
          >
            Public (Visible to everyone)
          </span>
        </div>
        <div
          aria-checked="false"
          aria-selected="true"
          class="selected item"
          role="option"
          style="pointer-events: all;"
        >
          <span
            class="text"
          >
            Private (Visible only to logged in users)
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
